src/main/java/org/opensearch/geospatial/ip2geo/common/URLDenyListChecker.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.ip2geo.common;
+
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.UnknownHostException;
+import java.util.List;
+
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.geospatial.annotation.VisibleForTesting;
+
+import inet.ipaddr.IPAddressString;
+
+/**
+ * A class to check url against a deny-list
+ */
+public class URLDenyListChecker {
+    private final ClusterSettings clusterSettings;
+
+    public URLDenyListChecker(final ClusterSettings clusterSettings) {
+        this.clusterSettings = clusterSettings;
+    }
+
+    /**
+     * Convert String to URL after verifying the url is not on a deny-list
+     *
+     * @param url value to validate and convert to URL
+     * @return value in URL type
+     * @throws UnknownHostException
+     * @throws MalformedURLException
+     */
+    public URL toUrlIfNotInDenyList(final String url) throws UnknownHostException, MalformedURLException {
+        return toUrlIfNotInDenyList(url, clusterSettings.get(Ip2GeoSettings.DATASOURCE_ENDPOINT_DENYLIST));
+    }
+
+    @VisibleForTesting
+    @SuppressForbidden(reason = "Need to connect to http endpoint to read GeoIP database file")
+    protected URL toUrlIfNotInDenyList(final String url, final List<String> denyList) throws UnknownHostException, MalformedURLException {
```

Comment:
why protected?

Reply:
For testing.

Reply:
Lets see if we can avoid this. We have been abusing this annotation for a long time. If we can remove this it will be great. if not please create a github issue to remove this.